File: 0150_evaluateReversePolishNotation/main.cpp
#include <vector>
#include <string> //stoi
#include <unordered_map>
#include <functional>
#include <stack>


class Solution {
public:
    int evalRPN(std::vector<std::string>& tokens) {
        std::unordered_map<std::string, std::function<int(int, int)>> map = {
            { "+" , [] (int a, int b) { return a + b; } },
            { "-" , [] (int a, int b) { return a - b; } },
            { "*" , [] (int a, int b) { return a * b; } },
            { "/" , [] (int a, int b) { return a / b; } }
        };
        std::stack<int> stack;
        for (std::string& s : tokens) {
            if (map.find(s) == map.end()) {
                stack.push(std::stoi(s));
            } else {
                int op1 = stack.top();
                stack.pop();
                int op2 = stack.top();
                stack.pop();
                stack.push(map[s](op2, op1));
            }
        }
        return stack.top();
    }
};

class Solution1 {
public:
    int evalRPN(std::vector<std::string>& tokens) {
        std::stack<int> s;
        for(int i = 0; i < tokens.size(); ++i){
            if(tokens[i] == "+" || tokens[i] == "-" || tokens[i] == "*" || tokens[i] == "/"){
                int op2=s.top();
                s.pop();
                int op1=s.top();
                s.pop();
                switch(tokens[i][0]){
                    case '+':
                        s.push(op1+op2);
                        break;
                    case '-':
                        s.push(op1-op2);
                        break;
                    case '*':
                        s.push(op1*op2);
                        break;
                    case '/':
                        s.push(op1/op2);
                        break;
                    default:
                        throw "operator is" + tokens[i] + "is not allowed";
                }
            }
            else{
                s.push(std::stoi(tokens[i]));
            }
        }
        return s.top();
    }
};

File: 0150_evaluateReversePolishNotation/main.py
from typing import List
import math

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for val in tokens:
            if val in {'+', '-', '*', '/'}:
                x2 = stack.pop()
                x1 = stack.pop()
                if val == '+':
                    y = x1 + x2
                elif val == '-':
                    y = x1 - x2
                elif val == '*':
                    y = x1 * x2
                elif val == '/':
                    y = math.trunc(x1 / x2)
                else:
                    raise Exception("not support" + val + "as an operator")
                stack.append(y)
            else:
                stack.append(int(val))
        return stack.pop()
